File: trie.py
print('trie')

class TrieNode:
    '''
    Trie Node
    '''

    def __init__(self, isValid=False, val=''):
        self.children = [None for _ in range(26)]
        self.valid = isValid

class Trie:
    '''
    Trie (prefix tree) implementation.
    Trie is a tree-like data structure whose nodes store the letters of an alphabet
    by structuring the nodes in a particular way, words and strings
    can be retrieved from the structure by traversing down a branch path of the tree.

    No node in the tree stores the key associated with that node;
    instead, its position in the tree defines the key with which it is associated.

    All the descendants of a node have a common prefix of the string associated with that node,
    and the root is associated with the empty string.

    Values are not necessarily associated with every node.

    Rather, values tend only to be associated with leaves, and with some inner nodes that correspond to keys of interest.
    '''

    def __init__(self):
        '''
        Initialize the Trie data structure.

        '''
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        '''
        Inserts a word into the trie.

        Time complexity: :math:`O(n)`, where n is the length of the word.

        :type word: str
        :rtype: None
        '''
        cur = self.root
        for ch in word:
            if cur.children[ord(ch)-97] is None:
                cur.children[ord(ch)-97] = TrieNode()
            cur = cur.children[ord(ch)-97]

        cur.valid = True


    def search(self, word: str) -> bool:
        '''
        Returns if the word is in the trie.

        Time complexity: :math:`O(n)`, where n is the length of the word.

        :type word: str
        :rtype: bool
        '''
        cur = self.root
        for ch in word:
            if cur.children[ord(ch)-97] is None:
                return False
            cur = cur.children[ord(ch)-97]

        return cur.valid


    def startsWith(self, prefix: str) -> bool:
        '''
        Returns if there is any word in the trie that starts with the given prefix.

        Time complexity: :math:`O(n)`, where n is the length of the prefix.

        :type prefix: str
        :rtype: bool
        '''
        cur = self.root
        for ch in prefix:
            if cur.children[ord(ch)-97] is None:
                return False
            cur = cur.children[ord(ch)-97]

        return True


# Your Trie object will be instantiated and called as such:
obj = Trie()
obj.insert("asdf")
obj.insert("as")
obj.insert("asdd")
obj.insert("asdd")
obj.insert("apple")
print(obj.search("as"))
print(obj.search("asdd"))
print(obj.search("asddd"))
print(obj.search("asdddd"))
print(obj.search("apple"))
print(obj.startsWith("app"))
print(obj.startsWith("a"))
print(obj.startsWith("as"))
# param_3 = obj.startsWith(prefix)

File: trie/trie.rs
use std::collections::HashMap;

#[derive(Default)]
struct TrieNode {
    children: HashMap<char, TrieNode>,
    is_end_of_word: bool,
}

#[derive(Default)]
pub struct Trie {
    root: TrieNode,
}

impl Trie {
    pub fn new() -> Self {
        Default::default()
    }

    pub fn insert(&mut self, word: &str) {
        let mut curr = &mut self.root;
        for ch in word.chars() {
            curr = curr.children.entry(ch).or_default();
        }
        curr.is_end_of_word = true;
    }

    pub fn search(&self, word: &str) -> bool {
        let mut curr = &self.root;
        for ch in word.chars() {
            match curr.children.get(&ch) {
                Some(node) => curr = node,
                None => return false,
            }
        }
        curr.is_end_of_word
    }

    pub fn starts_with(&self, prefix: &str) -> bool {
        let mut curr = &self.root;
        for ch in prefix.chars() {
            match curr.children.get(&ch) {
                Some(node) => curr = node,
                None => return false,
            }
        }
        true
    }
}

fn main() {
    let mut trie = Trie::new();
    trie.insert("apple");
    println!("{}", trie.search("apple"));   // true
    println!("{}", trie.search("app"));     // false
    println!("{}", trie.starts_with("app")); // true
    trie.insert("app");
    println!("{}", trie.search("app"));     // true
}
